src/main/java/com/exactpro/th2/service/AbstractGrpcService.java

```diff
@@ -125,11 +94,115 @@ protected S getStub(Message message, Map<String, String> properties) {
+                IllegalStateException exception = new IllegalStateException("Not yet init");
+                exception.addSuppressed(firstException);
+                throw exception;
+            }
+            RuntimeException retryRootException = new RuntimeException(ROOT_RETRY_SYNC_EXCEPTION_MESSAGE);
+            handleStatusRuntimeException(retryPolicy, retryRootException, firstException, 0);
+
+            for (int attempt = 1; attempt <= retryPolicy.getMaxAttempts(); attempt++) {
+                try {
+                    Thread.sleep(retryPolicy.getDelay(attempt - 1));
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    retryRootException.addSuppressed(e);
+                    throw retryRootException;
+                }
+
+                try {
+                    return retryCall.get();
+                } catch (StatusRuntimeException e) {
+                    handleStatusRuntimeException(retryPolicy, retryRootException, e, attempt);
+                } catch (Exception e) {
+                    retryRootException.addSuppressed(e);
+                    throw retryRootException;
+                }
+            }
+
+            throw retryRootException;
+        } catch (RuntimeException e) {
+            throw new RuntimeException(ROOT_RETRY_SYNC_EXCEPTION_MESSAGE, e);
+        }
+    }
+
+    private <T> T executeWithRetrySync(Supplier<T> call) {
+        return executeWithRetrySync(call, call, Function.identity()::apply);
+    }
+
+    private void handleStatusRuntimeException(@NotNull RetryPolicy retryPolicy,
+                                              @NotNull RuntimeException retryRootExp,
+                                              @NotNull StatusRuntimeException grpcExp,
+                                              int attempt) {
+        retryRootExp.addSuppressed(grpcExp);
+        Status.Code statusCode = grpcExp.getStatus().getCode();
+        switch (statusCode) {
+            case UNKNOWN:
+            case DEADLINE_EXCEEDED:
+                throw retryRootExp;
+            case CANCELLED:
+                if (retryPolicy.retryInterruptedTransaction()
+                        && STATUS_DESCRIPTION_OF_INTERRUPTED_REQUEST.equals(grpcExp.getStatus().getDescription())) {
+                    LOGGER.warn("GRPC blocking request has been interrupted during handle");
+                } else {
+                    throw retryRootExp;
+                }
+        }
+        LOGGER.warn("Can not send GRPC blocking request. Retrying. Current attempt = {}", attempt, grpcExp);
+    }
+    @NotThreadSafe
+    private class RetryIterator<T> implements Iterator<T> {
+        private final Supplier<Iterator<T>> method;
+        private Iterator<T> internal;
+        private boolean hasResponse = false;
+
+        public RetryIterator(Supplier<Iterator<T>> method) {
+            this.method = method;
+            this.internal = method.get();
```

Comment:
Just for my understanding: if we do a method call here and it throws an exception (no connection to the server for example) how do we retry after that? Or there are guarantees that the method call itself does not do anything - only the call to `hasNext` does the actual request?

I am asking that because right now I don't see that we do any retries in this case (maybe I simply don't see them)

Reply:
gRPC generate code which creates the next actions:
1) create BlockingResponseStream. this class implements Iterator<Response> interface, method.get() returns it
2) call asyncUnaryRequestCall and pass BlockingResponseStream.listener(). The listener buffers incoming events including exceptions
3) BlockingResponseStream pull events from internal buffer and process them when user call hasNext, next methods

Reply:
Thank you @Nikita-Smirnov-Exactpro for clarification. I believe that it would be good to have this information in the code to justify why we do what we do.